che-incubator/che-code
# [WIP] The PowerShell extension-related fixes

Signed-off-by: Artem Zatsarynnyi <azatsary@redhat.com>

### What does this PR do?
1. **Terminals:**
* Returns back the VS Code built-in terminal with all its features enabled (e.g. [Shell integration](https://code.visualstudio.com/docs/terminal/shell-integration), [contributed Terminal Profiles](https://code.visualstudio.com/docs/terminal/profiles), etc.).
* Contributes "opening a terminal to any DevWorkspace container" feature exclusively by the VS Code extension API, without the VS Code internals modifications.

2. **Tasks:**
* All the functionality related to running the Devfile Commands plugged into VS Code through the extensions.
The `che-commands` extension relies on the `che-terminal` extension's API for running the Devfile Commands in any container via the machine-exec server.
* Returns back the possibility of running any non-Che VS Code Tasks, but not Che Tasks (Devfile Commands) only.

3. Terminal-related changes removed obstacles that prevented the **PowerShell VS Code extension** from working properly.
_See more details in the comment https://github.com/eclipse/che/issues/21626#issuecomment-1219475237._

### What issues does this PR fix or reference?
- https://github.com/eclipse/che/issues/21673
- https://github.com/eclipse/che/issues/21626
- https://github.com/eclipse/che/issues/21466
- https://github.com/eclipse/che/issues/21616

### How to test this PR?

#### Test opening the terminals

##### Use case 1: there're multiple contributed containers
1. Start a DevWorkspace with the test project https://github.com/azatsarynnyy/java-spring-petclinic/tree/trm-ext-multiple
2. In the Command Palette, call the command `Terminal: Create New Terminal to DevWorkspace Container`.
3. VS Code **should** ask to choose a container as there're several contributed containers except for the editor's container.

##### Use case 2: there's one contributed container (except for the editor's container)
1. Start a DevWorkspace with the test project https://github.com/azatsarynnyy/java-spring-petclinic/tree/trm-ext 
2. In the Command Palette, call the command `Terminal: Create New Terminal to DevWorkspace Container`.
3. VS Code **should not** ask to choose a container as there's a single contributed container (`mariadb`) except the editor's container.
4. VS Code should open a terminal to the container `mariadb`.

##### Use case 3: there're no contributed containers (except for the editor's container)
In this case, the behavior should be exactly the same as opening a built-in VS Code terminal.
1. Start a DevWorkspace with the test project https://github.com/azatsarynnyy/java-spring-petclinic/tree/trm-ext-0 
2. In the Command Palette, call the command `Terminal: Create New Terminal to DevWorkspace Container`.
3. VS Code **should not** ask to choose a container as there're no contributed containers except for the editor's container.
4. VS Code should open a terminal to the editor container `tools`.
5. Check that the opened terminal **is a full-featured built-in terminal**, not the machine-exec terminal. They differ in that the machine-exec terminal doesn't support features like Shell integration.


#### Test running the tasks
1. Start a DevWorkspace with the test project https://github.com/azatsarynnyy/java-spring-petclinic/tree/trm-ext
2. Create test `tasks.json` with a `shell` task. E.g.: `Terminal > Configure Tasks... > Create tasks.json file from template > maven`.
3. Run the `shell` task.
4. The task should be executed as expected. It shouldn't hang as described in https://github.com/eclipse/che/issues/21466

#### Test PowerShell extension functionality
1. Start a DevWorkspace with the test project https://github.com/azatsarynnyy/PowerShellEditorServices/tree/trm-ext.
2. Install the recommended `PowerShell` extension.
3. Make sure a new Terminal Profile `pwsh` is contributed to the `Launch Profile...` menu.
4. After activating the extension (e.g. on opening a `*.ps1` file), a new terminal `PowerShell Extension` is launched automatically.
5. Check the editor autocompletion in the `*.ps1` file work as expected.
6. Check that the `PowerShell` left-side panel is populated with the commands.
7. Check that no problems are reported in the extension's logs (`View`>`Output`).
